test/evp_libctx_test.c

```diff
@@ -443,6 +443,190 @@ static void collect_cipher_names(EVP_CIPHER *cipher, void *cipher_names_list)
     sk_OPENSSL_CSTRING_push(names, EVP_CIPHER_name(cipher));
 }
 
+static int rsa_keygen(const char *alg, int bits, EVP_PKEY **pub, EVP_PKEY **priv)
+{
+    int ret = 0;
+    EVP_PKEY_CTX *keygen_ctx = NULL;
+    unsigned char *pub_der = NULL;
+    const unsigned char *pp = NULL;
+    long len = 0;
+
+    if (!TEST_ptr(keygen_ctx = EVP_PKEY_CTX_new_from_name(libctx, alg, NULL))
+        || !TEST_int_gt(EVP_PKEY_keygen_init(keygen_ctx), 0)
+        || !TEST_true(EVP_PKEY_CTX_set_rsa_keygen_bits(keygen_ctx, bits))
+        || !TEST_int_gt(EVP_PKEY_keygen(keygen_ctx, priv), 0)
+        || !TEST_int_gt(len = i2d_PublicKey(*priv, &pub_der), 0))
+        goto err;
+    pp = pub_der;
+    if (!TEST_ptr(d2i_PublicKey(EVP_PKEY_RSA, pub, &pp, len)))
+        goto err;
+    ret = 1;
+err:
+    OPENSSL_free(pub_der);
+    EVP_PKEY_CTX_free(keygen_ctx);
+    return ret;
+}
+
+static int kem_rsa_gen_recover(void)
+{
+    int ret = 0;
+    EVP_PKEY *pub = NULL;
+    EVP_PKEY *priv = NULL;
+    EVP_PKEY_CTX *sctx = NULL, *rctx = NULL;
+    unsigned char secret[256] = { 0, };
+    unsigned char ct[256] = { 0, };
+    unsigned char unwrap[256] = { 0, };
+    size_t ctlen = 0, unwraplen = 0, secretlen = 0;
+    const char *alg = "RSA";
+    const char *kem_type = "RSASVE";
+
+    if (!TEST_true(rsa_keygen(alg, 2048, &pub, &priv)))
+        goto done;
+
+    if (!TEST_ptr(sctx = EVP_PKEY_CTX_new_from_pkey(libctx, pub, NULL)))
+        goto done;
+
+    if (!TEST_int_eq(EVP_PKEY_encapsulate_init(sctx, kem_type), 1))
+        goto done;
```

Comment:
I know this is coming from

>> IMO, we should have the KEM mode as a required parameter for the encapsulate/decapsulate operations from start.
>
> Done..
> Fixed up docs + added more tests for boundary conditions.

and I hate to come out only now after you did the work, but while this solves most of my previous concerns, it also creates new ones: so far for most of the `EVP` API, you cal `EVP_verb` using (a ctx with) a key of a certain type and you get the implementation of that verb for the key type, if I write an app or lib that uses `EVP_sign_init()` I don't need to change my code to handle an RSA or an EC key.

Having the "mode name" as part of `EVP_PKEY_encapsulate_init()` means that this is not true for the new API: we go back to having branches depending on the key type.

For example, if there was an RFC for using KEM in TLS 1.3, in a patch for libssl we couldn't simply just call `EVP_derive()` for this bunch of protocol codepoints and call `EVP_encapsulate()` for this other bunch of protocol codepoints, but we would need to call `EVP_encapsulate_init()` with a different set of argument for each of the KEM codepoints.

Is this something we really want?

@levitte, as our resident historian, what do you think about this? Am I being overzealous again?

Reply:
I think this just creates more problems than it solves.
Embedding a kdf inside a particular kem can be done fairly easily..

Reply:
@romen, we've already had the signal more than once: stop inventing new patterns if there already is one.  In particular, for functions that use an EVP_PKEY (and an EVP_PKEY_CTX, which is pretty darn visible in this case), there's an already existing pattern `EVP_PKEY_verb`...  or you can go with the more global operation pattern `EVP_CamelVerb`.

> Having the "mode name" as part of `EVP_PKEY_encapsulate_init()` means that this is not true for the new API: we go back to having branches depending on the key type.

Hmmm, `EVP_PKEY` is hardly a mode name, it's a common prefix for functions that operate using an `EVP_PKEY`.  So if you see that as a "mode", then that "mode" is already indicated by the context type that's used.

Regarding what KDF to use internally in a specific KEM, isn't it just a matter of passing it the name of that KDF, just like we pass cipher names and digest names to a KDF?  Or have I misunderstood the problem?

Reply:
Sorry, I made a mess in the original post. It was not at all about naming or prefixes, but I see now that I managed to forget to put PKEY in so many places that my message got mixed.

My only concern is about passing the mode name, "RSASVE" in this case, during the encapsulate_init call.

For sign_init, derive_init, etc. you don't pass such an argument: whatever is the key you passed, you get the corresponding operation: the result is that the only thing changing is they key object and the code does not need branches to call sign, derive, etc. differently based on the key type.

When we will have post-quantum KEMs, that argument will make even less sense, as we will have one encapsulate/decapsulate operation for key type.

The modes standardized by FIPS for RSA in this are the exception, exposing their internals (e.g. pick this or that kdf, this or that symmetric wrapping, etc.).

If a key type offers tweakability w.r.t. operations it can support (e.g. I could chose A or B as the signature operation for a key of type FOO) shouldn't that be modeled as a key parameter rather than as an argument for encapsulate/decapsulate, the same way if the operation itself is tweakable (e.g. get a shared secret of length X or Y running the encapsulate with this key of type FOO+B) we set that as a parameter for the operation ctx?

Reply:
So you are saying you want it as a set_param instead...
Now that we have decided this is only an 'RSA' type that will now work..
However RSASVE will be the default if you dont set it - unless you think forcing the user to set it is a good idea.

Reply:
We could have encapsulate_init or decapsulate_init return -2 unless the param has been set if we want to force it.

But before you go and do it, what do others think?

Reply:
set_params() is normally called after the init().
And if you need to set it I am not sure we gain much by shifting it..
It could also be a NULL parameter for the init in the cases you are talking about.

Reply:
No, I propose to set RSASVE as a property of the key, not of the KEM CTX. If RSASVE is not set on the key, that key provides a "not supported" result if you try to initialize encapsulate or decapsulate.

This would allow the encapsulate_init of RSASVE or RSA-KEM-FOO to behave differently if needed, but still work on plain RSA keys, as long as the user explicitly set this key to work as RSASVE or RSA-KEM-FOO w.r.t. which KEM operation to implement.

If in 3.1 we then decide that it is fine to set RSA-KEM-FOO as the default KEM for plain RSA we just lift the restriction that returns -2 a less one of the valid KEM modes has been selected.

The goal is to have the branches about how to handle the key in a single place, when the key object is instantiated or right after it, and prevent spaghettification when the key is actually used.

For a future PQC-BAR KEM key that only does PQC-BAR as a KEM there will be no need to set the KEM_MODE property on the key, as that key only supports one KEM implementation.

Reply:
Basically the dispatch table for the KEM operation associated with RSA is set to a proxy.

The proxy returns -2 by default, or calls into the actual RSASVE or tomorrow RSA-KEM-FOO implementation if this key instance has been configured to run a specific KEM mode.

Reply:
So what do you think the call would look like here?

privkey = d2i_AutoPrivateKey_ex(NULL, &der, derlen, NULL, NULL);
privctx = EVP_PKEY_CTX_new_from_pkey(libctx, privkey, NULL);
EVP_PKEY_encapsulate_init(privctx);

Reply:
> So what do you think the call would look like here?
> 
>~~~C
> privkey = d2i_AutoPrivateKey_ex(NULL, &der, derlen, NULL, NULL);
> privctx = EVP_PKEY_CTX_new_from_pkey(libctx, privkey, NULL);
> EVP_PKEY_encapsulate_init(privctx);
>~~~

- As it is this code should fail for any known pkey type, as none of our internal key types support KEM by default.
- If the decoded `privkey` was a PQC KEM provided by an external provider, it would work without any code change
- if the decoded `privkey` is plain RSA and the application/protocol externally signalled that an `RSASVE` operation is needed, before line 2, the application/library would need to add:
  
  ~~~diff
    privkey = d2i_AutoPrivateKey_ex(NULL, &der, derlen, NULL, NULL);
  + EVP_PKEY_set_property(privkey, "KEM_MODE", "RSASVE"); /* this function does not exist currently, naming might also be wrong, this is just how I envision we can support that a signle key type might support more implementations of the same operation */
    privctx = EVP_PKEY_CTX_new_from_pkey(libctx, privkey, NULL);
    EVP_PKEY_encapsulate_init(privctx);
  ~~~

Reply:
I can't find the name right now, but I am pretty sure we have somewhere a function that operates on a `EVP_PKEY *` and that ultimately calls the `OSSL_FUNC_KEYMGMT_SET_PARAMS` for that key.

Reply:
No there isnt..

There are ctx setters (for operations) using EVP_PKEY_CTX_set_params()
And there are getters such as EVP_PKEY_get_octet_string_param() to retrieve things..

EVP_PKEY_fromdata() will load a key from raw parameters, but that is not what you want.

Reply:
Should we have it? How are the `keymgmt->settable_params` or `keymgmt->set_params` (and the `get` counterparts) reached? Are they only used by the core?

Reply:
I updated https://github.com/openssl/openssl/pull/12750#discussion_r487585439 to use an equivalent of `EVP_PKEY_CTX_set_params()` that calls the `keymgmt->set_params` instead of `${op}->set_params`.

Reply:
Yep and now we are heading into very confusing territory as to what is done during the ctx_set_params and what is allowed for set_params

Reply:
Well, missing accessors for `keymgmt->{settable_params,gettable_params,set_params,get_params}` qualify as a bug anyway, don't they?

`settable_params` lets the provider establish what is settable for a key object.

Here is an attempt at implementing the setter:

~~~diff
diff --git a/crypto/evp/p_lib.c b/crypto/evp/p_lib.c
index fd2a6c5abc..0179edc2f5 100644
--- a/crypto/evp/p_lib.c
+++ b/crypto/evp/p_lib.c
@@ -1320,6 +1320,37 @@ int EVP_PKEY_set1_tls_encodedpoint(EVP_PKEY *pkey,
     return 1;
 }
 
+/*
+ * Setter for OSSL_PARAMs for an provider-native EVP_PKEY object.
+ *
+ * Return 1 on success, 0 or negative for errors.
+ *
+ * In particular return -2 if any of the params is not supported.
+ */
+int EVP_PKEY_set_params(EVP_PKEY *pkey, const OSSL_PARAM *params)
+{
+    const OSSL_PARAM *p, *settable_params;
+
+    if (pkey == NULL || params == NULL)
+        return 0;
+
+    /*
+     * This operation is supported only for provider-native keys, return
+     * -2 if this is not the case for pkey.
+     */
+    if (pkey->ameth != NULL || pkey->keymgmt == NULL || pkey->keydata != NULL)
+        return -2;
+
+    settable_params = EVP_KEYMGMT_settable_params(keymgmt);
+    for (p = params; p->key != NULL; p++) {
+        /* Check the keymgmt actually supports this parameter */
+        if (OSSL_PARAM_locate_const(settable_params, p->key) == NULL )
+            return -2;
+    }
+
+    return evp_keymgmt_set_params(pkey->keymgmt, pkey->keydata, params);
+}
+
 size_t EVP_PKEY_get1_tls_encodedpoint(EVP_PKEY *pkey, unsigned char **ppt)
 {
     int rv;

~~~

The problem with doing it as a `EVP_PKEY_CTX_set_params`, is that it comes after the `init` has been called, and we can't assume that `encapsulate_init` for `RSASVE` and `RSA-KEM-FOO` are going to be identical.

That is why I see "what is the supported KEM operation for this key" a key property and not a `ctx` property.

Reply:
> No, I propose to set RSASVE as a property of the key

Why do you want to attach that to the key?

> The goal is to have the branches about how to handle the key in a single place, when the key object is instantiated or right after it, and prevent spaghettification when the key is actually used.

So basically, what you really want is for "RSA-SVE" to be seen as a KEM algorithm that uses RSA keys.

Unfortunately, what you're suggesting, setting what's essentially an *operation mode* (just like for example "CBC" is a mode for doing things like "AES") to the key material sends a different message, that there is an RSA-SVE key of sorts.  However, making such a thing leads us to a whole set of other problems, as has been said before, so I for one would prefer not to go down this route.
Besides...  what do you think should be done after the operation is performed?  Unset the KEM mode in the RSA key?

I admit that it would be nice if we had the same conceptual structure for modes with EVP_PKEY operations as we do with EVP_CIPHER, so we could sensibly specify that the "RSA-SVE" KEM algorithm should be used, the same way that we specify that "AES-256-CBC" should be used when we init a cipher operation.  It would allow us to specify the variants for doing KEM that you seem too desire, and I know that there were EC-related things that you saw a solution for that way as well.
*Unfortunately*, all public key operations so far are structured around that for any key type FOO, there is *one* signature implementation, *one* encryption implementation, *one*...  you get the picture.  So we end up with implementations that support different modes to be able to provide the variations.  This is visible all the way to the `openssl pkeyutl`, where it's all about operations being performed on a key, and any variations of those operations are determined with `-pkeyopt` (which is such a misnomer, as it's *operation* options rather than key options).
The *one* keytype, *one* operation implementation "principle" runs so fundamental in everything EVP_PKEY that I have difficulty seeing how we could change that without either causing an utter mess in commands like `openssl pkeyutl`.  I think that this is beyond what is sensible to do for 3.0.

So in summary, I would indeed suggest that the mode be passed as an operation parameter.  That would also make it easy to implement `openssl pkeyutl -inkey foo.pem -encapsulate` with `-pkeyopt mode:rsa-sve` to select what encapsulation variant should be used.

Reply:
> > No, I propose to set RSASVE as a property of the key
> 
> Why do you want to attach that to the key?

`encapsulate_init` for "RSASVE" and "RSA-KEM-FOO" might look very different:
- the provider-side context structures might require different fields, e.g. because internally the latter also needs to keep track of a KDF operation and of a symmetric encryption operation.

  This can be overcome by having a single provider-side ctx struct that contains a union of the fields needed for both operations, but is doing so pushing complexity that is due to the `libcrypto` legacy API onto the authors of providers. 
  From an external provider author this extra complexity in the provider API is not really justifiable, and we can't hide behind "we have to support legacy" as the provider API is brand new!
- the `encapsulate_init` or `decapsulate_init` to perform all the initialization stuff might very much require to know if this is "RSASVE" or "RSA-KEM-FOO", e.g. for the latter we might need to fetch and init a KDF implementation and a symmetric encryption implementation

> > The goal is to have the branches about how to handle the key in a single place, when the key object is instantiated or right after it, and prevent spaghettification when the key is actually used.
> 
> So basically, what you really want is for "RSA-SVE" to be seen as a KEM algorithm that uses RSA keys.

My original proposal was to have RSASVE as a separate key type, but that does not play well with encoders/decoders;

Given that instead we need to handle plain RSA keys and have them support alternative KEM operations, I am trying to figure out how to do that, before `encapsulate_init` or `decapsulate_init` are called.

Alternatively we can accept the current proposal of passing the name of the mode as an argument for `encapsulate_init` or `decapsulate_init`, but that also comes with its own drawbacks from an API design POV:
- why is the `encapsulate/decapsulate` the only one in the `EVP_PKEY` family that also asks for a `name` parameter?
- libssl and external applications will need to keep track of "what kind of RSA key is this? one that encapsuquacks like RSASVE or like RSA-KEM-FOO?" alongside the key for all its lifetime; in `libssl` for example we create the ephemeral key used during the TLS 1.3 key agreement in one function, and than we operate on it calling `derive()` somewhere else; when we will update `libssl` to support KEM, we will have to parse the codepoint lookup table to decide what is the `name` argument to pass to `decapsulate_init` and add complexity compared to the current state of things were all the differences among the various `EVP_PKEY` options are handled at keygen time once and for all.

> Unfortunately, what you're suggesting, setting what's essentially an _operation mode_ (just like for example "CBC" is a mode for doing things like "AES") to the key material sends a different message, that there is an RSA-SVE key of sorts. However, making such a thing leads us to a whole set of other problems, as has been said before, so I for one would prefer not to go down this route.
> Besides... what do you think should be done after the operation is performed? Unset the KEM mode in the RSA key?

In my mental model, if this plain RSA key is created to support RSASVE as a KEM operation, it supports it until it is freed.

> _Unfortunately_, all public key operations so far are structured around that for any key type FOO, there is _one_ signature implementation, _one_ encryption implementation, _one_... you get the picture. So we end up with implementations that support different modes to be able to provide the variations. This is visible all the way to the `openssl pkeyutl`, where it's all about operations being performed on a key, and any variations of those operations are determined with `-pkeyopt` (which is such a misnomer, as it's _operation_ options rather than key options).
> The _one_ keytype, _one_ operation implementation "principle" runs so fundamental in everything EVP_PKEY that I have difficulty seeing how we could change that without either causing an utter mess in commands like `openssl pkeyutl`. I think that this is beyond what is sensible to do for 3.0.

This is the original reason why I thought `RSASVE` should be its own key type, seeing it from the provider POV, and let encoder/decoder handle the complexity of casting the plain RSA key to an RSASVE key, based on the other information available at decoding time, coming from the protocol or the application, that specifies how the decoded key should be used.

In the end:
- we have a FIPS document that says here is a cryptosystem called "RSASVE" that operates on "plain RSA" keys;
- we have a previous revision of the same document that offered two other encapsulate/decapsulate cryptosystems that also operate on "plain RSA" keys;
- we have no widely adopted standard that says "RSASVE" is THE encapsulate/decapsulate operation for "plain RSA" keys, so what is to stop IETF or ISO to release a spec that becomes more popular than the current revision of the FIPS document we are looking at and that sets the de-facto standard for KEM over RSA to "RSA-KEM-BAR"? In this case it's not even about conflicts, FIPS is not mandating to ONLY use "RSASVE" as an encapsulate/decapsulate primitive for plain RSA keys. So it really is just a matter of which one will be the most widely adopted standard that implements an actual protocol using a KEM-like operation on plain RSA keys.

How can we reconciliate this with our "one encoding"-"one keytype"-"one implementation for each operation"?

> So in summary, I would indeed suggest that the mode be passed as an operation parameter. That would also make it easy to implement `openssl pkeyutl -inkey foo.pem -encapsulate` with `-pkeyopt mode:rsa-sve` to select what encapsulation variant should be used.

This could also be possible if:
- we had a `-pkeykeyopt` argument for `pkeyutl`, to call `EVP_PKEY_ctrl_str()` on the decoded key object
- we had a `EVP_PKEY_ctrl_str()` equivalent of `EVP_PKEY_CTX_ctrl_str()`, mapping on `EVP_PKEY_set_param()`

Reply:
> This is the original reason why I thought `RSASVE` should be its own key type, seeing it from the provider POV, and let encoder/decoder handle the complexity of casting the plain RSA key to an RSASVE key, based on the other information available at decoding time, coming from the protocol or the application, that specifies how the decoded key should be used.

When I wrote this I meant to say that, given the inherent ambiguity of selecting one among a few alternatives of encapsulate/decapsulate operations that operate on plain RSA keys, either the protocol/application encodes the RSA key with a specific OID for RSASVE/RSA-KEM-FOO, or it signals with some codepoint on the wire how to use the RSA key, or it implicitly declares in its spec that a plain RSA key must be used using "BAR" as the associated KEM-like operation.

This information should thus always be available at the moment in which the plain RSA key is being decoded, or immediately around it: if RSASVE was a separate key type we could have a param passed to the ctx used for the decode operation that pick RSASVE as the keymgmt to use rather than plain RSA.

Reply:
[sigh] I'm obviously not getting the point through that you're creating a *world* of complications that I believe is unsuitable for 3.0.

So how about an imagined call sequence instead.  This is one added line to what @slontis asked, and all it means is that `EVP_PKEY_encapsulate_init()` won't be able to determine the KEM mode, but `EVP_PKEY_encapsulate()` will:

``` C
privkey = d2i_AutoPrivateKey_ex(NULL, &der, derlen, NULL, NULL);
privctx = EVP_PKEY_CTX_new_from_pkey(libctx, privkey, NULL);
EVP_PKEY_encapsulate_init(privctx);
EVP_PKEY_CTX_set_encapsulate_mode(privctx, "RSA-SVE");  /* example name */
```

Reply:
EVP_PKEY_CTX_set_encapsulate_mode() can fail also..

This is as good as you can do with the current model which is all setup to work off a ctx.
Doing anything else would just be a shift away from how everything else is working currently.
The ctx stuff is already confusing enough as it is without coming up with yet another mechanism.

Reply:
> [sigh] I'm obviously not getting the point through that you're creating a _world_ of complications that I believe is unsuitable for 3.0.

I am actually receptive to that point and I opened the discussion saying that maybe adding RSASVE in 3.0 was not the best idea:

>> @romen [said](https://github.com/openssl/openssl/pull/12750#issuecomment-684760703)
>> I believe it would be better to drop support for this in 3.0 […]
>
> @slontis [said](https://github.com/openssl/openssl/pull/12750#issuecomment-685171372)
> As previously stated - it is a fundamental requirement of SP800-56B - so it needs to go in.

RSASVE appeared on the horizon well after alpha6, or I would have discussed these points way earlier!

> So how about an imagined call sequence instead. This is one added line to what @slontis asked, and all it means is that `EVP_PKEY_encapsulate_init()` won't be able to determine the KEM mode, but `EVP_PKEY_encapsulate()` will:
> 
> ```c
> privkey = d2i_AutoPrivateKey_ex(NULL, &der, derlen, NULL, NULL);
> privctx = EVP_PKEY_CTX_new_from_pkey(libctx, privkey, NULL);
> EVP_PKEY_encapsulate_init(privctx);
> EVP_PKEY_CTX_set_encapsulate_mode(privctx, "RSA-SVE");  /* example name */
> ```

It would work, unless "RSA-SVE" and "RSA-KEM-FOO" have different requirements to run `encapsulate_init()` (e.g. fetching and initializing KDF and symmetric crypto): in that case you need to know "RSA-SVE" or "RSA-KEM-FOO" before calling `encapsulate_init()`.

Why is this one line change worse?

```c
privkey = d2i_AutoPrivateKey_ex(NULL, &der, derlen, NULL, NULL);
EVP_PKEY_set_encapsulate_mode(privkey, "RSA-SVE");  /* example name */
privctx = EVP_PKEY_CTX_new_from_pkey(libctx, privkey, NULL);
EVP_PKEY_encapsulate_init(privctx);
```

or

```c
privkey = d2i_AutoPrivateKey_ex(NULL, &der, derlen, NULL, NULL);
EVP_PKEY_set_alias_type(privkey, EVP_PKEY_RSASVE); /*similar to SM2 */
privctx = EVP_PKEY_CTX_new_from_pkey(libctx, privkey, NULL);
EVP_PKEY_encapsulate_init(privctx);
```

Reply:
> It would work, unless "RSA-SVE" and "RSA-KEM-FOO" have different requirements to run `encapsulate_init()` (e.g. fetching and initializing KDF and symmetric crypto): in that case you need to know "RSA-SVE" or "RSA-KEM-FOO" before calling `encapsulate_init()`.

So...  why *must* that be done in `encapsulate_init()`?  Look at other operations where we do this kind of thing; they handle the fetching as part of parameter setting.

Reply:
> > It would work, unless "RSA-SVE" and "RSA-KEM-FOO" have different requirements to run `encapsulate_init()` (e.g. fetching and initializing KDF and symmetric crypto): in that case you need to know "RSA-SVE" or "RSA-KEM-FOO" before calling `encapsulate_init()`.
> 
> So... why _must_ that be done in `encapsulate_init()`? Look at other operations where we do this kind of thing; they handle the fetching as part of parameter setting.

_Now_ I get what both you and @slontis meant with

> EVP_PKEY_CTX_set_encapsulate_mode() can fail also..

Ok, sorry, I wish I understood this point earlier and saved everyone my walls of text!

If that is what we are already doing elsewhere — i.e. shallow `whatever_init()` and then the "real" initialization happening once all the required params have been set or while the required params are being set — I guess it could work also in this case.

Sidenote: now that I realize we have this pattern elsewhere... how are we making sure that the order of `EVP_PKEY_CTX_set_params()` calls does not become a potentially breaking-change factor?
It seems like a very dangerous and slippery pattern to adopt!
Do we have extensive tests, testing all possible combinations whenever there is more than one param to set on a given `ctx` to actually trigger the initialization?

Reply:
> Do we have extensive tests, testing all possible combinations

I think you already know the answer to that one..

Reply:
> > Do we have extensive tests, testing all possible combinations
> 
> I think you already know the answer to that one..

Is it one of the following? :smile: 

- A) "we only used that pattern when there was only one param required to trigger the actual initialization"
- B) "we documented that all the ctx params required for initialization must be passed all at once, in the same `EVP_PKEY_CTX_set_params()` call, in a well-defined order, immediately after `EVP_whatever_init()`"
- C) "This…. is…. Spartaaaaaaaaaa" :crossed_swords: :hole:

Reply:
I'm not sure why the different extra parameters must be given in a certain order..... The doer `encapsulate()` would simply have to check that all the data it has makes sense together, and fail if not. How many tests do you need for that?

Reply:
> I'm not sure why the different extra parameters must be given in a certain order..... The doer `encapsulate()` would simply have to check that all the data it has makes sense together, and fail if not. How many tests do you need for that?

So the actual `init()` step does not happen when `EVP_PKEY_CTX_set_params()` is pushing the required `init()` parameters, but when the actual operation is executed?
- this means that we can do only minimal validation when `EVP_PKEY_CTX_set_params()` is called, and errors are triggered away from the problematic line that provided invalid params
- previously lightweight operations, that only executed the actual operation, now need to do also all the costly fetches and checks that in 1.1.1 were relegated to the `whatever_init()` function: this might bite us as soon as we and the external users start looking at benchmark results and compare 1.1.1 with 3.0; some performance degradation is expected due to the new architecture that adds more indirection and string parsing/caching but in `speed` that would have been hidden/minimized as `whatever_init()` is called once, and then `whatever()` is called in a tight loop; if `whatever()` does the heavy lifting of the initialization all the slowdowns are going to pile up instead!


If instead the bulk of the initialization (and validation of the parameters) happens during `EVP_PKEY_CTX_set_params()`, the risk is that we test:

```C
OSSL_PARAM params[] = { FOO_PARAM_MODE, FOO_PARAM_A, FOO_PARAM_B, OSSL_PARAM_END };
EVP_foo_init(pctx);
EVP_PKEY_CTX_set_params(pctx, params);
EVP_foo(pctx);
```

Our test might pass, but it might fail if instead we had 3 separate `EVP_PKEY_CTX_set_params()` calls, or if the order of things inside `params` was changed.

That would make the order of function calls part of the things covered by our stability promise, and one more quirk that our users would need to learn (the hard way, because it is difficult to debug!).

Reply:
Why can't the fetching happen during set_params? That's how we do it elsewhere, and in that case, the doer only has to do som pretty quick checks, such as checking that it has a non-NULL EVP_KDF pointer, for example. That's also very much like the 1.1.1 way, except a method pointer would be passed instead of a name.

Reply:
I am not saying that it is impossible to write it right, I am just noting that this pattern has a higher risk of being implemented in the wrong way, if not today, tomorrow, and that is where testing the various ordering combinations would be useful, or testing that we fail exactly in the way we documented if a wrong combination of params is provided or if a necessary param is omitted, etc.

Anyway, this is beside the point for this PR: `EVP_PKEY_CTX_set_params()` after `operation_init()` has been called is the pattern we used elsewhere, and should work also here.

I was only surprised as I wasn't aware this was happening, and it seems one of those patterns that they would recommend against in SWENG courses, especially for the maintenance and testing costs associated with it.

Reply:
@romen, there are certainly parts of the API that could do with a redesign.  That is, however, an effort for another time.